### PR TITLE
Refine arguments for error interceptors

### PR DESCRIPTION
**Pull request by xamgore**
Any is not really a good type for error handling. I believe they should be Request / Response objects.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -120,7 +120,7 @@ export interface CancelTokenSource {
 }
 
 export interface AxiosInterceptorManager<V> {
-  use(onFulfilled?: (value: V) => V | Promise<V>, onRejected?: (error: any) => any): number;
+  use(onFulfilled?: (value: V) => V | Promise<V>, onRejected?: (error: V) => any): number;
```
Comment (Alanscut) [inline index.d.ts]: A successful callback should return V（AxiosResponse）, but when there is an error, maybe it should return `AxiosError` insteadof `V`?

**Comment by Alanscut** [comment]
Actually I think there will be no problem even if I don’t change it.

File: index.d.ts (hunk 1 above)
Comment (xamgore) [reply index.d.ts]: Are you sure about `AxiosError`?

**Comment by chinesedfan** [comment]
Sorry, we don't know which type the error is. Because it can be thrown from anywhere.

**Comment by xamgore** [comment]
@chinesedfan can we improve typing by writing `AxiosError | any`? IDE would improve their intellisense significantly.

**Comment by chinesedfan** [comment]
@xamgore You can use a type assertion(https://github.com/axios/axios/issues/2013#issuecomment-468172172) or contribute a type guard(#1863). In fact, I like type guard better.

**Comment by xamgore** [comment]
Type assertion requires a hidden knowledge of `AxiosError` type existence. Same with a type guard — you have to read about it in docs or the source code before usage.

On the other hand if `onRejected` accepts a union, people will be able to use optional chaining, and IDE will enumerate its inner fields:

```diff
export interface AxiosInterceptorManager<V> {
-  use(onFulfilled?: (value: V) => V | Promise<V>, onRejected?: (error: V) => any): number;
+  use(onFulfilled?: (value: V) => V | Promise<V>, onRejected?: (error: AxiosError | any) => any): number;
  eject(id: number): void;
}
```

**Comment by chinesedfan** [comment]
@xamgore `AxiosError | any` looks making sense, though it seems to be weird at first glance. Can you find some documents or example links about it? Then we can persuade more people.

**Comment by xamgore** [comment]
Typescript documentation [has a similar case](https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types): 

They define two interfaces, possibly intersecting:

```ts
interface Bird {
    fly();
    layEggs();
}

interface Fish {
    swim();
    layEggs();
}

function getSmallPet(): Fish | Bird {
    // ...
}

let pet = getSmallPet();
pet.layEggs(); // okay
pet.swim();    // errors

pet?.layEggs(); // okay
pet?.swim(); // okay, though swim() wasn't called
```

Then they provide a user type-guard:
```
function isFish(pet: Fish | Bird): pet is Fish {
    return (pet as Fish).swim !== undefined;
}

if (isFish(pet)) pet.swim() // ok
```

In our case `Fish | Bird` === `AxiosError | any`.

**Comment by chinesedfan** [comment]
I know that official example. And just thought `any` as a bit special. Thanks for your work and I will do some further investigating. Also waiting for others' opinions in #2729.

**Comment by ahuglajbclajep** [comment]
Using `asserts x is T` added in [TypeScript 3.7](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions), you might be able to write [as follows](http://www.typescriptlang.org/play/?ssl=11&ssc=27&pln=11&pc=1#code/MYGwhgzhAECCUFMBOAXAlgewHYFElIyWgQA8UEsATGPAogbwF8AoNLcpAMzGAThMwQASgggAHbBAQAeACoA+aPWiUwKMAC5osgNzQWbDt179BtQnMWlyVGvkJLoSURKxSA-FtgCMwl5JkFPRZmTgBXLGB0bGhIKVQASQhvM3skSwAKZCQtMCwATwBKXMRUGGzoNBgU33N0hSVmaErOaCz8ADoqmog6wsbm5pQACwIAd2gsBAn4eOjcNIyAIgA5DBRYrFNatIBCJcKdJv1mEJR8sT4AETUwaABeRxItLDCAWwAjZH0jygRQMDOaDASQbbK5ApHFBIfIDWKlFBJHp1aQ3dTydpIQ7HEFuDAgBAdEAYADmmI6znEAXcHVU6g6JEO0AA9MziPgvD5emlUbd5KdgWpgMM4biIPjCcSydkmaz2TlNvlTkA).

```ts
function assertIsAxiosError<T>(err: any): asserts err is AxiosError<T> {
  if (err.isAxiosError) {
    throw new AssertionError("Not an AxiosError!");
  }
}

// in onRejected function
// type Data = { x: number };
try {
  assertIsAxiosError<Data>(err);
  console.log(err.response?.data.x); // err: AxiosError<Data>
} catch {
  console.log(err); // err: any
}
```

I would be happy if this information is helpful for you.

**Pull request closed without merging**